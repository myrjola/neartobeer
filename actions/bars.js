import checkStatus from '../util/checkStatus';

export const REQUEST_BARS = 'REQUEST_BARS';
export const RECEIVE_BARS = 'RECEIVE_BARS';
export const ERROR_BARS = 'ERROR_BARS';
export const INVALIDATE_BARS = 'INVALIDATE_BARS';
export const SELECT_BAR = 'SELECT_BAR';
export const DESELECT_BAR = 'DESELECT_BAR';

export const invalidateBars = {
  type: INVALIDATE_BARS,
};

export const requestBars = {
  type: REQUEST_BARS,
};

export const receiveBars = json => ({
  type: RECEIVE_BARS,
  bars: json,
  receivedAt: Date.now(),
});

export const errorBars = error => ({
  type: ERROR_BARS,
  error,
});

export const selectBar = barId => ({
  type: SELECT_BAR,
  barId,
});

export const deselectBar = {
  type: DESELECT_BAR,
};

export function fetchBars() {
  return (dispatch) => {
    dispatch(requestBars);
    return fetch('http://beersunderthirty.org/places/places.json')
      .then(checkStatus)
      .then(response => response.json())
      .then(json => dispatch(receiveBars(json)))
      .catch(error => dispatch(errorBars(error)));
  };
}

function shouldFetchBars(state) {
  if (!state.bars) {
    return true;
  } else if (state.bars.isFetching) {
    return false;
  }

  // The cache invalidates after one hour.
  const ONE_HOUR = 60 * 60 * 1000; /* ms */
  if (Date.now() - state.bars.lastUpdated > ONE_HOUR) {
    return true;
  }

  return state.bars.didInvalidate;
}

export function fetchBarsIfNeeded() {
  return (dispatch, getState) => {
    if (shouldFetchBars(getState())) {
      return dispatch(fetchBars());
    }
  };
}

// Local Variables:
// mode: js2-jsx
// End:
