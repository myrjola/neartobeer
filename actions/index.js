import checkStatus from '../util/checkStatus';

export const SET_BEER_MAX_PRICE_CATEGORY = 'SET_BEER_MAX_PRICE_CATEGORY';
export const REQUEST_BARS = 'REQUEST_BARS';
export const RECEIVE_BARS = 'RECEIVE_BARS';
export const INVALIDATE_BARS = 'INVALIDATE_BARS';
export const REQUEST_WALKING_DIRECTIONS = 'REQUEST_WALKING_DIRECTIONS';
export const RECEIVE_WALKING_DIRECTIONS = 'RECEIVE_WALKING_DIRECTIONS';
export const ERROR_WALKING_DIRECTIONS = 'ERROR_WALKING_DIRECTIONS';

export const setMaxBeerPriceCategory = category => ({
  type: SET_BEER_MAX_PRICE_CATEGORY,
  category,
});

export const invalidateBars = {
  type: INVALIDATE_BARS,
};

export const requestBars = {
  type: REQUEST_BARS,
};

export const receiveBars = json => ({
  type: RECEIVE_BARS,
  bars: json,
  receivedAt: Date.now(),
});

function fetchBars() {
  return (dispatch) => {
    dispatch(requestBars);
    // TODO: How to handle errors? Remember to set state.bars.isFetching to false.
    return fetch('http://beersunderthirty.org/places/places.json')
      .then(checkStatus)
      .then(response => response.json())
      .then(json => dispatch(receiveBars(json)));
  };
}

function shouldFetchBars(state) {
  if (!state.bars.items) {
    return true;
  } else if (state.bars.isFetching) {
    return false;
  }
  return state.bars.didInvalidate;
}

export function fetchBarsIfNeeded() {
  return (dispatch, getState) => {
    if (shouldFetchBars(getState())) {
      return dispatch(fetchBars());
    }
  };
}

export const requestWalkingDirections = destination => ({
  type: REQUEST_WALKING_DIRECTIONS,
  destination,
});

export const receiveWalkingDirections = (origin, coords, destination) => ({
  type: RECEIVE_WALKING_DIRECTIONS,
  origin,
  coords,
  destination,
});

// transforms something like this geocFltrhVvDsEtA}ApSsVrDaEvAcBSYOS_@... to an array of coordinates
// TODO: make a readable one using the https://developers.google.com/maps/documentation/directions/intro.
const decodeGoogleDirectionsPolyline = function(t,e){for(var n,o,u=0,l=0,r=0,d= [],h=0,i=0,a=null,c=Math.pow(10,e||5);u<t.length;){a=null,h=0,i=0;do a=t.charCodeAt(u++)-63,i|=(31&a)<<h,h+=5;while(a>=32);n=1&i?~(i>>1):i>>1,h=i=0;do a=t.charCodeAt(u++)-63,i|=(31&a)<<h,h+=5;while(a>=32);o=1&i?~(i>>1):i>>1,l+=n,r+=o,d.push([l/c,r/c])}return d=d.map(function(t){return{latitude:t[0],longitude:t[1]}})};

export const errorWalkingDirections = error => ({
  type: ERROR_WALKING_DIRECTIONS,
  error,
});

function coordinatesMatch(coord1, coord2) {
  return coord1.latitude === coord2.latitude && coord1.longitude === coord2.longitude;
}

function coordinatesToString(coord) {
  return `${coord.latitude},${coord.longitude}`;
}

function fetchWalkingDirections(destination) {
  return (dispatch) => {
    dispatch(requestWalkingDirections(destination));
    navigator.geolocation.getCurrentPosition(
      (position) => {
        const origin = position.coords;
        // FIXME: Figure out how not to repeat ourselves. This key is also in AndroidManifests.xml.
        const APIKEY = '***REMOVED***';
        const url = `https://maps.googleapis.com/maps/api/directions/json?origin=${coordinatesToString(origin)}` +
              `&destination=${coordinatesToString(destination)}&key=${APIKEY}&mode=walking`;
        return fetch(url)
          .then(checkStatus)
          .then(response => response.json())
          .then(json => decodeGoogleDirectionsPolyline(json.routes[0].overview_polyline.points))
          .then(coords => dispatch(receiveWalkingDirections(origin, coords, destination)))
          .catch(error => dispatch(errorWalkingDirections(error)));
      },
      error => dispatch(errorWalkingDirections(error)),
    );
  };
}

function shouldFetchWalkingDirections(state, destination) {
  const destinationHasChanged = coordinatesMatch(destination, state.walkingDirections.destination);
  if (!state.walkingDirections) {
    return true;
  } else if (!destinationHasChanged && state.walkingDirections.isFetching) {
    return false;
  }
  return true;
}

export function fetchWalkingDirectionsIfNeeded(destination) {
  return (dispatch, getState) => {
    if (shouldFetchWalkingDirections(getState(), destination)) {
      return dispatch(fetchWalkingDirections(destination));
    }
  };
}

// Local Variables:
// mode: js2-jsx
// End:
