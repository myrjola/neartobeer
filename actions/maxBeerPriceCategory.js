export const SET_BEER_MAX_PRICE_CATEGORY = 'SET_BEER_MAX_PRICE_CATEGORY';

export const setMaxBeerPriceCategory = category => ({
  type: SET_BEER_MAX_PRICE_CATEGORY,
  category,
});

// LOCAL Variables:
// mode: js2-jsx
// End:
