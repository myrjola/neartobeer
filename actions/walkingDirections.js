import polyline from 'google-polyline';

import checkStatus from '../util/checkStatus';

export const REQUEST_WALKING_DIRECTIONS = 'REQUEST_WALKING_DIRECTIONS';
export const RECEIVE_WALKING_DIRECTIONS = 'RECEIVE_WALKING_DIRECTIONS';
export const ERROR_WALKING_DIRECTIONS = 'ERROR_WALKING_DIRECTIONS';

export const requestWalkingDirections = destination => ({
  type: REQUEST_WALKING_DIRECTIONS,
  destination,
});

export const receiveWalkingDirections = (origin, coords, destination) => ({
  type: RECEIVE_WALKING_DIRECTIONS,
  origin,
  coords,
  destination,
});

function decodeGoogleDirectionsPolyline(polylineString) {
  const coordinates = polyline.decode(polylineString);
  return coordinates.map(coordinate => ({ latitude: coordinate[0], longitude: coordinate[1] }));
}

export const errorWalkingDirections = error => ({
  type: ERROR_WALKING_DIRECTIONS,
  error,
});

export function coordinatesMatch(coord1, coord2) {
  return coord1.latitude === coord2.latitude && coord1.longitude === coord2.longitude;
}

function coordinatesToString(coord) {
  return `${coord.latitude},${coord.longitude}`;
}

function fetchWalkingDirections(destination) {
  return (dispatch) => {
    dispatch(requestWalkingDirections(destination));
    navigator.geolocation.getCurrentPosition(
      (position) => {
        const origin = position.coords;
        // FIXME: Figure out how not to repeat ourselves. This key is also in AndroidManifests.xml.
        const APIKEY = '***REMOVED***';
        const url = `https://maps.googleapis.com/maps/api/directions/json?origin=${coordinatesToString(origin)}` +
              `&destination=${coordinatesToString(destination)}&key=${APIKEY}&mode=walking`;
        return fetch(url)
          .then(checkStatus)
          .then(response => response.json())
          .then(json => decodeGoogleDirectionsPolyline(json.routes[0].overview_polyline.points))
          .then(coords => dispatch(receiveWalkingDirections(origin, coords, destination)))
          .catch(error => dispatch(errorWalkingDirections(error)));
      },
      error => dispatch(errorWalkingDirections(error)),
    );
  };
}

function shouldFetchWalkingDirections(state, destination) {
  const destinationHasChanged = coordinatesMatch(destination, state.walkingDirections.destination);
  if (!state.walkingDirections) {
    return true;
  } else if (!destinationHasChanged && state.walkingDirections.isFetching) {
    return false;
  }
  return true;
}

export function fetchWalkingDirectionsIfNeeded(destination) {
  return (dispatch, getState) => {
    if (shouldFetchWalkingDirections(getState(), destination)) {
      return dispatch(fetchWalkingDirections(destination));
    }
  };
}


// Local Variables:
// mode: js2-jsx
// End:
