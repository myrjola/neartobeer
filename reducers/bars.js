import { REQUEST_BARS, RECEIVE_BARS, INVALIDATE_BARS } from '../actions';

const BARS = [
  {
    post_id: 38,
    post_title: 'MacLarens',
    post_author: 1,
    post_content: 'Serves chilled Norrlands Guld on tap for only 29:-.' +
      'One of the most best places in town for cheap beer lovers.',
    post_category: 'Beer under 30',
    default_category: 'Beer under 30',
    post_tags: '',
    post_type: 'gd_place',
    post_status: 'publish',
    is_featured: '',
    geodir_video: '',
    post_address: 'Norrtullsgatan 11',
    post_city: 'Stockholm',
    post_region: 'Stockholm County',
    post_country: 'Sweden',
    post_zip: '113 29',
    post_latitude: 59.3419634,
    post_longitude: 18.051719899999966,
    geodir_timing: '',
    geodir_contact: '',
    geodir_email: '',
    geodir_website: '',
    geodir_twitter: '',
    geodir_facebook: '',
    geodir_special_offers: '',
    IMAGE: '',
  },
  {
    post_id: 39,
    post_title: 'Lion Bar Sveavägen 39',
    post_author: 1,
    post_content: 'The famous Lion Bar chain is a perfect find for bargain brew lovers! 7 Places all around' +
      ' Stockholm serve chilled Spendrups (40cl, tapped) for 23:- until 20h. After this the prices will increase ' +
      'in two steps: until midnight the price will be 36:-, thereafter 40:-. Go and get there early to get yourself' +
      ' a crispy brew for under 30:- !',
    post_category: 'Beer under 50',
    default_category: 'Beer under 30',
    post_tags: '',
    post_type: 'gd_place',
    post_status: 'publish',
    is_featured: '',
    geodir_video: '',
    post_address: 'Sveavägen 39 Stockholm',
    post_city: 'Stockholm',
    post_region: 'Stockholm County',
    post_country: 'Sweden',
    post_zip: '',
    post_latitude: 59.3404626,
    post_longitude: 18.05888660000005,
    geodir_timing: '',
    geodir_contact: '',
    geodir_email: '',
    geodir_website: '',
    geodir_twitter: '',
    geodir_facebook: '',
    geodir_special_offers: '',
    IMAGE: '',
  },
];

function bars(state = {
  isFetching: false,
  didInvalidate: false,
  items: BARS,
}, action) {
  switch (action.type) {
    case INVALIDATE_BARS:
      return Object.assign({}, state, {
        didInvalidate: true,
      });
    case REQUEST_BARS:
      return Object.assign({}, state, {
        isFetching: true,
        didInvalidate: false,
      });
    case RECEIVE_BARS:
      return Object.assign({}, state, {
        isFetching: false,
        didInvalidate: false,
        items: action.bars,
        lastUpdated: action.receivedAt,
      });
    default:
      return state;
  }
}

export default bars;

// Local Variables:
// mode: js2-jsx
// End:
