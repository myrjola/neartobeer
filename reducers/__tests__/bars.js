import mockStore from 'redux-mock-store';

import { requestBars, receiveBars, invalidateBars, fetchBarsIfNeeded } from '../../actions/bars';
import bars, { initialState } from '../bars';

const exampleBars = [
  {
    post_id: 38,
    post_title: 'MacLarens',
    post_author: 1,
    post_content: 'Serves chilled Norrlands Guld on tap for only 29:-.' +
      'One of the most best places in town for cheap beer lovers.',
    post_category: 'Beer under 30',
    default_category: 'Beer under 30',
    post_type: 'gd_place',
    post_status: 'publish',
    post_address: 'Norrtullsgatan 11',
    post_city: 'Stockholm',
    post_region: 'Stockholm County',
    post_country: 'Sweden',
    post_zip: '113 29',
    post_latitude: 59.3419634,
    post_longitude: 18.051719899999966,
  },
  {
    post_id: 39,
    post_title: 'Lion Bar Sveavägen 39',
    post_author: 1,
    post_content: 'The famous Lion Bar chain is a perfect find for bargain brew lovers! 7 Places all around' +
      ' Stockholm serve chilled Spendrups (40cl, tapped) for 23:- until 20h. After this the prices will increase ' +
      'in two steps: until midnight the price will be 36:-, thereafter 40:-. Go and get there early to get yourself' +
      ' a crispy brew for under 30:- !',
    post_category: 'Beer under 50',
    default_category: 'Beer under 30',
    post_type: 'gd_place',
    post_status: 'publish',
    post_address: 'Sveavägen 39 Stockholm',
    post_city: 'Stockholm',
    post_region: 'Stockholm County',
    post_country: 'Sweden',
    post_latitude: 59.3404626,
    post_longitude: 18.05888660000005,
  },
];

var store = null;

beforeEach(() => {
  store = mockStore({ bars: initialState });
  global.Date.now = jest.fn(() => 1482363367071);
});

it('invalidates content', () => {
  expect(bars(initialState, invalidateBars)).toMatchSnapshot();
});

it('requests new bars', () => {
  expect(bars(initialState, requestBars)).toMatchSnapshot();
});

it('receives new bars', () => {
  expect(bars(initialState, receiveBars(exampleBars))).toMatchSnapshot();
});

it('errors when HTTP request fails', async () => {
  fetch.mockResponseFailure('Service unavailable!');
  await store.dispatch(fetchBarsIfNeeded());
  expect(store.getActions()).toMatchSnapshot();
});

it('requests and receives new bars', async () => {
  fetch.mockResponseSuccess(JSON.stringify(exampleBars));
  await store.dispatch(fetchBarsIfNeeded());
  expect(store.getActions()).toMatchSnapshot();
});

it("doesn't request bars when fetch in progress", async () => {
  store = mockStore({ bars: bars(initialState, requestBars) });
  await store.dispatch(fetchBarsIfNeeded());
  expect(store.getActions()).toMatchSnapshot();
});

it("doesn't request bars when cache not invalidated", async () => {
  store = mockStore({ bars: bars(initialState, receiveBars(exampleBars)) });
  await store.dispatch(fetchBarsIfNeeded());
  expect(store.getActions()).toMatchSnapshot();
});

it('invalidates over one hour old cache', async () => {
  var receiveBarsState = { bars: bars(initialState, receiveBars(exampleBars)) };
  const ONE_HOUR = 60 * 60 * 1000; /* ms */
  receiveBarsState.bars.lastUpdated = Date.now() - ONE_HOUR - 1;
  store = mockStore(receiveBarsState);
  fetch.mockResponseSuccess(JSON.stringify([]));
  await store.dispatch(fetchBarsIfNeeded());
  expect(store.getActions()).toMatchSnapshot();
});

// Local Variables:
// mode: js2-jsx
// End:
