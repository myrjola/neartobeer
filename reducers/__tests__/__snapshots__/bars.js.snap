// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`doesn't request bars when cache not invalidated 1`] = `Array []`;

exports[`doesn't request bars when fetch in progress 1`] = `Array []`;

exports[`errors when HTTP request fails 1`] = `
Array [
  Object {
    "type": "REQUEST_BARS",
  },
  Object {
    "error": "Service unavailable!",
    "type": "ERROR_BARS",
  },
]
`;

exports[`invalidates content 1`] = `
Object {
  "didInvalidate": true,
  "isFetching": false,
  "items": Array [],
}
`;

exports[`receives new bars 1`] = `
Object {
  "didInvalidate": false,
  "isFetching": false,
  "items": Array [
    Object {
      "default_category": "Beer under 30",
      "post_address": "Norrtullsgatan 11",
      "post_author": 1,
      "post_category": "Beer under 30",
      "post_city": "Stockholm",
      "post_content": "Serves chilled Norrlands Guld on tap for only 29:-.One of the most best places in town for cheap beer lovers.",
      "post_country": "Sweden",
      "post_id": 38,
      "post_latitude": 59.3419634,
      "post_longitude": 18.051719899999966,
      "post_region": "Stockholm County",
      "post_status": "publish",
      "post_title": "MacLarens",
      "post_type": "gd_place",
      "post_zip": "113 29",
    },
    Object {
      "default_category": "Beer under 30",
      "post_address": "Sveav채gen 39 Stockholm",
      "post_author": 1,
      "post_category": "Beer under 50",
      "post_city": "Stockholm",
      "post_content": "The famous Lion Bar chain is a perfect find for bargain brew lovers! 7 Places all around Stockholm serve chilled Spendrups (40cl, tapped) for 23:- until 20h. After this the prices will increase in two steps: until midnight the price will be 36:-, thereafter 40:-. Go and get there early to get yourself a crispy brew for under 30:- !",
      "post_country": "Sweden",
      "post_id": 39,
      "post_latitude": 59.3404626,
      "post_longitude": 18.05888660000005,
      "post_region": "Stockholm County",
      "post_status": "publish",
      "post_title": "Lion Bar Sveav채gen 39",
      "post_type": "gd_place",
    },
  ],
  "lastUpdated": 1482363367071,
}
`;

exports[`requests and receives new bars 1`] = `
Array [
  Object {
    "type": "REQUEST_BARS",
  },
  Object {
    "bars": Array [
      Object {
        "default_category": "Beer under 30",
        "post_address": "Norrtullsgatan 11",
        "post_author": 1,
        "post_category": "Beer under 30",
        "post_city": "Stockholm",
        "post_content": "Serves chilled Norrlands Guld on tap for only 29:-.One of the most best places in town for cheap beer lovers.",
        "post_country": "Sweden",
        "post_id": 38,
        "post_latitude": 59.3419634,
        "post_longitude": 18.051719899999966,
        "post_region": "Stockholm County",
        "post_status": "publish",
        "post_title": "MacLarens",
        "post_type": "gd_place",
        "post_zip": "113 29",
      },
      Object {
        "default_category": "Beer under 30",
        "post_address": "Sveav채gen 39 Stockholm",
        "post_author": 1,
        "post_category": "Beer under 50",
        "post_city": "Stockholm",
        "post_content": "The famous Lion Bar chain is a perfect find for bargain brew lovers! 7 Places all around Stockholm serve chilled Spendrups (40cl, tapped) for 23:- until 20h. After this the prices will increase in two steps: until midnight the price will be 36:-, thereafter 40:-. Go and get there early to get yourself a crispy brew for under 30:- !",
        "post_country": "Sweden",
        "post_id": 39,
        "post_latitude": 59.3404626,
        "post_longitude": 18.05888660000005,
        "post_region": "Stockholm County",
        "post_status": "publish",
        "post_title": "Lion Bar Sveav채gen 39",
        "post_type": "gd_place",
      },
    ],
    "receivedAt": 1482363367071,
    "type": "RECEIVE_BARS",
  },
]
`;

exports[`requests new bars 1`] = `
Object {
  "didInvalidate": true,
  "isFetching": true,
  "items": Array [],
}
`;
